lib/AST/ASTContext.cpp

```diff
@@ -3697,6 +3697,26 @@ Type AnyFunctionType::composeInput(ASTContext &ctx, ArrayRef<Param> params,
   return TupleType::get(elements, ctx);
 }
 
+bool AnyFunctionType::equalParams(ArrayRef<AnyFunctionType::Param> a,
+                                  ArrayRef<AnyFunctionType::Param> b) {
+  if (a.size() != b.size())
+    return false;
+
+  for (unsigned i = 0, n = a.size(); i != n; ++i) {
+    if (a[i] != b[i])
+      return false;
+  }
+
+  return true;
+}
+
+bool AnyFunctionType::equalParams(ASTContext &ctx,
+                                  ArrayRef<AnyFunctionType::Param> a,
+                                  Type composedB) {
+  auto composedA = AnyFunctionType::composeInput(ctx, a, false);
+  return composedA->isEqual(composedB);
```

Comment:
This seems problematic. If composedB has a paren around a tuple, we'll end up stripping the paren off.

Reply:
Yeah, `isEqual` canonicalizes the types.